docs/context-helpers.md

```diff
@@ -0,0 +1,362 @@
+
+```
+{#login}
+  Welcome!
+  {:usernameError}
+  Your username was not found.
+  {:passwordError}
+  Your password was incorrect. You have {.} attempts remaining.
+  {:else}
+  Please log in!
+{/login}
+```
+```
+{
+  "login": function(chunk, context, bodies) {
+    var username = context.get("username"),
+        password = context.get("password"),
+        status = authorizeUser(username, password);
+
+    switch(status.message) {
+      case "OK":
+        return true;
+      case "InvalidUserName":
+        return chunk.render(bodies.usernameError, context);
+      case "InvalidPassword":
+        return chunk.render(bodies.passwordError,
+                            context.push(status.loginAttemptsRemaining);
+    }
+
+    return false;
+  }
+}
+```
+
+#### Params
+
+Context helpers can be passed parameters just like regular Dust sections. They are accessed through the **params** argument.
+
+```
+{#price value="39.9" /}
+```
+```
+{
+  "price": function(chunk, context, bodies, params) {
+    return chunk.write("$" + Number(params.value).toFixed(2));
+  }
+}
+```
+
+##### Evaluating a parameter
+
+If a parameter contains a Dust reference, you must evaluate the reference if you want to use it in your context helper. Reference evaluation is done using `dust.helpers.tap()` (provided as part of the `dustjs-helpers` addon).
+
+```
+{#say text="Hello {name}!"/}
+```
+```
+{
+  "say": function(chunk, context, bodies, params) {
+    var text = dust.helpers.tap(params.text, chunk, context);
+    text = text.toUpperCase();
+    return chunk.write(text);
+  }
+}
+```
+
+## Asynchronous context helpers
+
+Dust's asynchronous nature is one of its defining features. Writing context helpers in an async way lets you make HTTP requests or call services without blocking the rendering of your template.
+
+```
+{#ip}Your IP address: {ip}{/ip}
+```
+
+```
+{
+  "ip": function(chunk, context, bodies, params) {
+    return chunk.map(function(chunk) {
+      $.get("//ip.jsontest.com/", function(data) {
+        // data contains { "ip": "123.45.67.89" }
+        return chunk.render(bodies.block, context.push(data))
+                    .end();
+      });
+    });
+  }
+}
+```
+To start an asynchronous block, call `chunk.map`. Inside its callback function, you can perform any sync or async operations. The only difference is that when you're done, you must call `chunk.end` to signal that the async operations have completed.
+
+You can't return a value from an asynchronous helper like you can a normal one. You must return a chunk that has been `end`ed.
+
+## Try it out
+
+Try these exercises to help you further your understanding of context helpers!
+
+### The Bad API
+
+Your JSON data is badly-formatted, as seen in the sample. Write a helper to reformat this data to make the template work.
```

Comment:
I didn't do the web component formatting yet because I wasn't sure if we had it working. The sample is just the JSON data (minus the solution)